### DISCUSSION TITLE
Mfront / Abaqus interface in windows environment

### DISCUSSION THREAD

**Discussion by A-Bourceret** [thelfer/tfel#47] [Q&A] [answered]
Dear all,
I’m struggling with compilation problems to interface Mfront with Abaqus. The calculation is run using a bat file, and I modify the custom_v6.env to add compilators for Abaqus. 
Following the guidelines of the documentation I modify my custom_v6.env to add compilators for Abaqus. For that I use the windows gives file equivalent to the Linux description in the Mfront documentation, the file win86_64.env. But it seems that some modification as to be done for the C++ compiler options. In the .log file I can see that the Microsoft compilator is found by Abaqus, but compilation problems at the beginning of the umat.cpp file are encountered. 

Best regards,

Alexandre Bourceret

**Comment by fpguillet** [comment, marked as answer]
Several points.
1. You probably haven't run the MSVC init script (paths, etc). You can either run this batch file or, if installed, open a Visual Studio (the MS IDE + environment not the editor) command window (cmd prompt, it's available from the windows menu). That's  probably why the include files are not found. 
2. Even then, MSVC will complain at link stage because it doesn't allow static variables in a library that doesn't have a main entry point, at least not in the format of object code file ABAQUS is using. So the solution is to add umat.cpp in the MFront generated library, libABAQUSBehaviour.DLL, and use a very simple, dummy umat.cpp that calls this DLL.
3. The link flags in the .env file must be modified to take into account the DLL, also you need to modifiy the library path so that at runtime ABAQUS can find and call the DLL.

I can help with 2 & 3, I suppose you'll find answers for point 1 depending on what you have installed (compiler tools only, Visual Studio, etc.).

Regards,
Francois.

**Reply by A-Bourceret** [reply]
Hi François,

Thanks for your precise answer  ! The first point is achieve, and as you guess I'm now dealing with a link problem.
For the two other steps I might need your help, because I don't how to call .dll using a .cpp file. I try to look at some examples but it's seems not trivial. I don't know either how to modify the flag of the .env file. I change the path usub_lib_dir wich point the ABAQUSBEHAVIOUR.dll directory but it seems not to work.
[File.zip](https://github.com/thelfer/tfel/files/7628425/File.zip)

Thank you anyway for this great summary of the tasks ! 
Best regards,
Alexandre

**Reply by fpguillet** [reply]
Alexandre,

(make sure to read this answer fully before jumping to code/project editing. If you can reuse my DLL and umat it will  be much more simple.)

First you have to generate your cpp source files using MFront and the ABAQUS interface. Personally I compiled TFEL using Visual Studio 2019 Community edition. 3.4 compiles with 2 minor tweaks on MSVC include files.
The ABAQUSBEHAVIOUR dll is compiled using the generated project, no modification. That's the easy part.
Now you have to encapsulate umat.cpp in a 'bridge' DLL. First, edit the umat.cpp file and replace the __cplusplus macro by _MSVC_LANG macro to check against C++ supported version. __cplusplus does not work properly when using MSVC (sigh). You also have to rename the umat functions since they will be called by another umat and no longer directly from ABAQUS. I used the name "umat_mfront" but you can choose any other name. I think there is not other modification, in any case all the files are in the archive (look for abaqus folder, umat_or.cpp is the original file generated by MFront, umat.cpp deeper in the folder tree ("DLLtampon") is after modifications). Then compile a DLL from umat.cpp. You can use the CMake project I include, you will see it's very simple.
Now you have to copy both DLLs, .exp and .lib files to the abaqus input file location (actually I 'm sure you don't need all of those apart from the DLLs and umatdll.lib but better be safe).
Now you need a simple umat.cpp file to call the 'umat_mfront' function in the original umat.cpp file which has been modified and packed in a DLL. Look at the MFrontElas2 archive which contains the whole ABAQUS folder (inp files, etc). The simple umat is called 'umat_th.cpp'. It just passes on the call from ABAQUS to the bridge DLL verbatim. This allows the use of static variables in umatdll.dll whereas the original umat.cpp can't be compiled as is by ABAQUS command line.
Last step is ensuring libraries are found both at compile time (for compiling umat.cpp) and runtime (for running ABAQUS). You have to add the ABAQUS project folder to your current PATH (I use PowerShell for that) and modify ABAQUS link command. This second step is conveniently done using a file 'abaqus_v6.env' placed along the ABAQUS input file and which is automatically used when compiling the umat. Again, all these files are in the MFrontElas3 archive. I don't think any other modification for ABAQUS environment commands/flags is necessary but I've been through so many tests I can't guarantee that.

The test case I include with this post runs ok but after that you're on your own. The next step is probably to check that everything is ok when considering two different MFront materials in the same input file.

Once you've been through all that, you can reuse the umatdll.dll and umat_th.cpp for any other MFront project, actually. In fact you can probably drop in these files (umatdll.* and umat_th.cpp) from the archive I provide (MFrontElas3.zip), put your own ABAQUSBEHAVIOUS.DLL and ABAQUS project and it *should* work. 

François.
[MFrontElas3.zip](https://github.com/thelfer/tfel/files/7631708/MFrontElas3.zip)
[Elastique3.zip](https://github.com/thelfer/tfel/files/7631725/Elastique3.zip)
